src/infrastructure/content/ContentCommunicator.java

```diff
@@ -1,24 +1,257 @@
+	 */
+	private static HashMap<String, String> imageMap = new HashMap<String, String>();
+	
+	/**
+	 * Creating a HashMap to store the identifier and handler of subscriber of this module,
+	 * identifier is of type String while handler is of type IContentNotificationHandler.
+	 */
+	private static HashMap<String, IContentNotificationHandler> handlerMap = new HashMap<String, IContentNotificationHandler>();
+	
+	/**
+	 * An instance of ServerPort class is fetched and stored in the variable serverPort.
+	 */
+	private IServerPort serverPort = ContentFactory.getServerPort();
+	/**
+	 * The variable port will store the port number of the board which the client is accessing.
+	 */
+	private int port;
+	
+	/**
+	 * This variable will store the communicator of networking module to send data over the network.
+	 */
+	private ICommunicator communicator = CommunicatorFactory.getCommunicator(0);
+	
+	/**
+	 * logger is the instance of the class which implements ILogger interface.
+	 */
+	private static ILogger logger = LoggerFactory.getLoggerInstance();
+	
+	/**
+	 * This variable will store the String which will be passed to log method of logger
+	 */
+	private static String logMessage;
+	
+	/**
+	 * Constructor. Inside this, an instance of NetworkMessageHandler class is created and then,
+	 * it is used for subscribing to the networking module to receive messages related to content.
+	 */
+	protected ContentCommunicator() {
+		INotificationHandler nmh = new NetworkMessageHandler();
+		communicator.subscribeForNotifications("content", nmh);
+	}
+	
+	/**
+	 * This method provides userName to other classes within this package.
+	 * @return userName - username of the client.
+	 */
+	protected static String getUserName() {
+		if (userName != null) {
+			logMessage = "getUserName method of ContentCommunicator class is executing";
+			logger.log(ModuleID.INFRASTRUCTURE, LogLevel.INFO, logMessage);
+		}
+		else {
+			logMessage = "value of userName yet to be set in ContentCommunicator class";
+			logger.log(ModuleID.INFRASTRUCTURE, LogLevel.WARNING, logMessage);
+		}
+		return userName;
+	}
+	
+	/**
+	 * This method provides HashMap imageMap to other classes within this package.
+	 * @return imageMap - The HashMap which stores the userName and userImage.
+	 */
+	protected static HashMap<String, String> getImageMap() {
+		logMessage = "getImageMap method of ContentCommunicator class is executing";
+		logger.log(ModuleID.INFRASTRUCTURE, LogLevel.INFO, logMessage);
+		return imageMap;
+	}
+	
+	/**
+	 * This method allows other classes within this package to set the HashMap imageMap.
+	 * @param imagemap - The HashMap which stores the userName and userImage of all clients.
+	 */
+	protected static void setImageMap(HashMap<String, String> imagemap) {
+		logMessage = "setImageMap method of ContentCommunicator class is executing";
+		logger.log(ModuleID.INFRASTRUCTURE, LogLevel.INFO, logMessage);
+		imageMap = imagemap;

```

Comment:
Deep copy vs Shallow copy problem arises in line 115. `imageMap` data member becomes a reference to the input `imagemap` and not its copy that you intended through the set function.

If `imagemap` gets deleted, `imageMap` will also become null.

To copy a map into another map object, use `putAll()` or copy constructor of `HashMap` class.
Reference - https://www.geeksforgeeks.org/how-to-clone-a-map-in-java/

Reply:
resolved